### ISSUE TITLE
Allow failed image/receipt uploads to be retried

### ISSUE THREAD

**Issue by justinpersaud** [Expensify/App#50507]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** Any -- latest version
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @justinpersaud 
**Slack conversation:** https://expensify.slack.com/archives/C05LX9D6E07/p1728408442408079

## Action Performed:
Break down in numbered steps

Unclear of the exact way to reproduce this, but you can upload a receipt via Submit Expense to another user and then somehow disrupt the upload either by losing internet connection or some other means.

## Expected Result:
The receipt should fail to upload, but there should be a retry button

## Actual Result:
The error asks the user to download the image and then upload it again which is not a great UX

## Workaround:
Yes

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
![image](https://github.com/user-attachments/assets/a4c4a7b7-396d-4bd0-85ee-edd70e475f0f)

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by justinpersaud** [comment]
This is my first time creating a new feature request here so let me know if I'm missing something. I don't know that a project is necessary for improving UX here, is it?

**Comment by dannymcclain** [comment]
I don't really think so—but maybe that depends on how we solve the problem. I feel like we could just tweak the error message and add a retry link in the message.

Maybe something like this? (bold = link)

>There was an error uploading your receipt. **Try again.** Alternatively, **download the file** or dismiss this error to delete it.

**Comment by JmillsExpensify** [comment]
I like tweaking the error message. Let's start there

**Comment by justinpersaud** [comment]
It was raised in slack that the word download actually doesn't make much sense here because when the error happens, the file was not yet uploaded. The file is actually being fetched from the user's device.

**Comment by dannymcclain** [comment]
>It was raised in slack that the word download actually doesn't make much sense here because when the error happens, the file was not yet uploaded. The file is actually being fetched from the user's device.

So what happens when a user clicks that link currently? I'm sure we can tweak the error message to make sense if we understand what's actually happening.

**Comment by justinpersaud** [comment]
I believe it fetches the image from the user device. 

Downloading seems to imply that it's coming from an external source, like our servers or the Internet somewhere, when in reality, it's still on the user's device and just needs to be retried.

**Comment by dannymcclain** [comment]
Could we have it say:

>There was an error uploading your receipt. **Try again**. Alternatively, **save the file** or dismiss this error to delete it.

?

**Comment by dannymcclain** [comment]
Hmm yeah I could see that—we might want to get some other's to weigh in too. But if we wanted to go that route we could just use our standard confirmation modal.

**Comment by JmillsExpensify** [comment]
I tend to think one confirmation is enough, though I'll defer to y'all.

**Comment by justinpersaud** [comment]
I'm fine with that. I think the important part here is just getting the retry button in place to upload again.

@sakluger what are the next steps here? I'm new to this process but happy to help from the internal engineering side.

**Comment by sakluger** [comment]
@justinpersaud can this be handled externally or does it need to be done by an internal engineer?

- If it can be done externally, we can just add the `external` label, which will open up the issue for contributors to propose solutions.
- If it requires a backend engineer, then we apply the `internal` label and you (or someone else) can make the necessary PR

Either way, let's make sure the expected behavior in the OP is accurate. I updated it, can you review what I wrote up there?

**Comment by justinpersaud** [comment]
looks good to me, and I believe this is external because it's all happening with the application, not the backend.

**Comment by sakluger** [comment]
No proposals yet.

**Comment by dannymcclain** [comment]
Still waiting for proposals.

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Allow failed image/receipt uploads to be retried

### What is the root cause of that problem?
New Feature - adding 

### What changes do you think we should make in order to solve the problem?

We should add a `Try again` link that will dismiss the error and at the same time navigate the user to receipt step. By the way, although the local file blob link is available in the error Row we can not access local files because for security reasons we cannot access local files via JS browser; the user specifically need to select the file so we will navigate to the receipt step.
We will add retry param here and navigate to the receipt selector if it is true at end of the function
https://github.com/Expensify/App/blob/c3dde46202d2f48934ef7aa617d69c8b4c0b5ff6/src/components/ReportActionItem/MoneyRequestView.tsx#L476
```
                        onClose={(retry = false) => {

...

if (retry) {
                                Navigation.navigate(
                                    ROUTES.MONEY_REQUEST_STEP_SCAN.getRoute(
                                        CONST.IOU.ACTION.EDIT,
                                        iouType,
                                        transaction?.transactionID ?? '-1',
                                        report?.reportID ?? '-1',
                                        Navigation.getReportRHPActiveRoute(),
                                    ),
                                );
                            }
```
pass onClose here
https://github.com/Expensify/App/blob/c3dde46202d2f48934ef7aa617d69c8b4c0b5ff6/src/components/MessagesRow.tsx#L47

And inside DotIndicator we will change the copy to an appropriate text and 
https://github.com/Expensify/App/blob/c3dde46202d2f48934ef7aa617d69c8b4c0b5ff6/src/components/DotIndicatorMessage.tsx#L64-L74
add `Try again` Text Link and on `onPress` we will call onClose passing `true` to the retry param

### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
Here it is @dukenv0307 

https://github.com/user-attachments/assets/2fe3d910-a2f3-4357-9cee-b9afde41ebf6

**Comment by dannymcclain** [comment]
>I think we want to upload the failed image immediately instead of going to Scan page

Yeah I agree. I imagined this retrying "in place"—so not taking the user anywhere. cc @Expensify/design for viz and any additional thoughts.

**Comment by justinpersaud** [comment]
I agree. Ideally the retry button just repeats the last action to upload.

@FitseTLT are you saying this is not possible because of security constraints within iOS now?

**Comment by FitseTLT** [comment]
> I agree. Ideally the retry button just repeats the last action to upload.
> 
> @FitseTLT are you saying this is not possible because of security constraints within iOS now?

Not IOS, only disallowed in web. Read https://stackoverflow.com/questions/64250573/access-local-filesystem-from-web-browser-react/64250626

**Comment by shawnborton** [comment]
Yeah Danny that would be my vote as well.

**Comment by shawnborton** [comment]
I think the idea is that it would try to either download the file to your computer or save the file to your device (add it to the photo gallery)

**Comment by FitseTLT** [comment]
> @FitseTLT How does the `Download the file` link work?
> 
> <img alt="Screenshot 2024-10-31 at 10 05 39" width="469" src="https://private-user-images.githubusercontent.com/129500732/381808701-1d2aeb06-81df-4b32-81b3-9fc5a572d204.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzNzE1NzcsIm5iZiI6MTczMDM3MTI3NywicGF0aCI6Ii8xMjk1MDA3MzIvMzgxODA4NzAxLTFkMmFlYjA2LTgxZGYtNGIzMi04MWIzLTlmYzVhNTcyZDIwNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAzMVQxMDQxMTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNjdjZmNlYjc3ZjJmMDRiNGQyNGQ4MzNjZjRmZTFkY2M0NDA0OGQ4YTQyZDYwNzU0YzU5NjJjZmFlY2M2MGZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fZN917ZfTkaBI57jCSEPcoft5FIiuLdrmgKdZxsK1mk">

Via fetch for web @dukenv0307

**Comment by FitseTLT** [comment]
> @FitseTLT Can you update the proposal to fit the expectation above? Thanks

That would need separate treatment for the 4 platforms. I will look into it when the price gets bumped 👍

**Comment by sakluger** [comment]
@FitseTLT that sounds fair. We've bumped the price to $500.

**Comment by FitseTLT** [comment]
> @FitseTLT that sounds fair. We've bumped the price to $500.

Yep

**Comment by sakluger** [comment]
Not overdue - we just got a new proposal, let's see what @dukenv0307 thinks of it!

**Comment by FitseTLT** [comment]
[Updated](https://github.com/Expensify/App/issues/50507#issuecomment-2445393084)

**Comment by FitseTLT** [comment]
@dukenv0307 Have you got a chance to review my [proposal](https://github.com/Expensify/App/issues/50507#issuecomment-2445393084) which works for all platform. Also I have uploaded a POC video with the test branch. Thx!

**Comment by MariaHCD** [comment]
I see @justinpersaud is already assigned as the internal engineer here, I don't think there's need for another :) Un-assigning myself but let me know if I'm needed here.

**Comment by justinpersaud** [comment]
@nyomanjyotisa proposal sounds good to me too

**Comment by justinpersaud** [comment]
@JmillsExpensify I recall you said not to bother about the confirmation to dismiss the error here, is that still true?

**Comment by sakluger** [comment]
If we do want to keep it, then I think the copy should read:

> Once you dismiss the error, you will no longer be able to recover the receipt. <Dismiss / Cancel>

But I think it's probably fine to just remove the dismissal confirmation.

**Comment by trjExpensify** [comment]
Ehhh, people get pretty touchy about us losing their receipts. We've seen it already. I think we keep the secondary modal personally.

**Comment by sakluger** [comment]
@nyomanjyotisa sounds like we're keeping the dismissal confirmation. Can you use my [proposed copy above](https://github.com/Expensify/App/issues/50507#issuecomment-2463399372)?

**Comment by sakluger** [comment]
We have a [draft PR](https://github.com/Expensify/App/pull/52294).

**Comment by sakluger** [comment]
Sure thing! @nyomanjyotisa or @dukenv0307, once you start a discussion on Slack, can you please link it here as well so I can track the progress of the issue? Thanks!

**Comment by sakluger** [comment]
This is still being discussed in Slack.

**Comment by sakluger** [comment]
I think we're close to wrapping up the Slack discussion.

**Comment by sakluger** [comment]
Welp, one more decision left to make in that Slack thread. We are discussing whether to show an error message when offline. This depends on whether the file can be successfully downloaded while offline.

**Comment by sakluger** [comment]
We're back to work! @nyomanjyotisa has everything he needs to continue working on the PR.

**Comment by sakluger** [comment]
@nyomanjyotisa any updates on the PR?

**Comment by sakluger** [comment]
Thank you!

**Comment by sakluger** [comment]
Looks like we're close based on the PR! Test builds are generated so QA can test.

**Comment by sakluger** [comment]
Hey @dukenv0307, could you please complete the BZ checklist?

**Comment by sakluger** [comment]
Summarizing payment on this issue:

Contributor: @nyomanjyotisa $500, paid via Upwork
Contributor+: @dukenv0307 $500, will pay via Upwork once BZ checklist is completed

**Comment by sakluger** [comment]
Thanks! All done.
